### DISCUSSION TITLE
Best way to index a `Box<dyn noodles::noodles_sam::alignment::record::Sequence>`

### DISCUSSION THREAD

**Discussion by snalty** [zaeleus/noodles#256] [Q&A] [answered]
Hello,

I was wondering if anyone knew the best way to index this type with a `Range<Position>`? I cannot index it normally like I can the RecordBuf sequence type. I can do 

```
let sequence = noodles::sam::alignment::record_buf::Sequence::from(record.sequence().iter().collect::<Vec<u8>>());
```

and index that but that seems inefficient.

Any ideas? :)

**Comment by zaeleus** [comment, marked as answer]
I would recommend chaining [`Iterator::skip`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.skip) and [`Iterator::take`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.take). Depending on what you want to do with it, you can either iterate over the bases or copy into a new list, e.g.,

```rust
// cargo add bstr@1.9.1
// cargo add noodles@0.70.0 --features core,sam

use bstr::ByteSlice;
use noodles::{
    core::Position,
    sam::{self, alignment::record::Sequence},
};

fn main() -> Result<(), noodles::core::position::TryFromIntError> {
    let sequence_buf = sam::alignment::record_buf::Sequence::from(b"ACGTNTGCA");
    let sequence: Box<dyn Sequence> = Box::new(&sequence_buf);

    // 1-based: 123456789
    //          ACGTNTGCA
    //           ^^^^
    let start = Position::try_from(2)?;
    let end = Position::try_from(5)?;

    let s = usize::from(start) - 1;
    let e = usize::from(end) - 1;
    let len = e - s + 1;

    let subsequence: Vec<u8> = sequence.iter().skip(s).take(len).collect();

    dbg!(subsequence.as_bstr());
    // => "CGTN"

    Ok(())
}
```

**Reply by snalty** [reply]
This is perfect, thank you very much. I have successfully implemented this now and it is working well!

**Comment by mbhall88** [comment]
I have some follow up questions relating to this.

I have a trait `Mismatches` that I had originally implemented on [`noodles_sam::Record`](https://docs.rs/noodles-sam/latest/noodles_sam/record/struct.Record.html) - i.e., the concrete `struct`. However, I realised it is best to generalise this so I now implement my `Mismatches` trait on [`noodles_sam::alignment::Record`](https://docs.rs/noodles-sam/latest/noodles_sam/alignment/record/trait.Record.html). For context, I am iterating over alignment records using `noodles_util::alignment`, so each record is essentially `Box<dyn noodles_sam::alignment::Record>`.

The problem I have run into when making this trait implementation on the `Record` trait is that where previously I was retrieving a base in the sequence like this `self.sequence().as_ref()[idx]` (where `idx` is `usize`), I now have to do `self.sequence().iter().nth(idx).ok_or(anyhow::anyhow!("No base found"))`.

Indexing into the sequence on the concrete struct is obviously `O(1)`, but indexing into the sequence on the `Record` trait is `O(n)`, where `n` is the length of the sequence. Are you able to explain why this is so? Is it something along the lines of the trait is "lazy" and never actually reads in the sequence, whereas the struct does, so in the end they have the same time complexity?

And of course, am I going about this all wrong?

PS. your support on this repository is incredible and I have learnt so many things about Rust in my interactions with you (and reading your responses to others). So a HUGE thank you.

**Reply by zaeleus** [reply]
> Are you able to explain why this is so? 

The original question asked how to get a substring of the sequence. This isn't possible in all implementations of [`Sequence`](https://docs.rs/noodles/0.78.0/noodles/sam/alignment/record/trait.Sequence.html) because the underlying representation of the data may not be able to be sliced, e.g., a BAM record sequence that is 4-bit encoded.

We could add a method signature to the trait that returns a single base with a given index. This is now tracked in #283.
